Binary Distribution: Relocate RPATH on Cray

eugeneswalker opened the pull request:
Do RPATH relocation on Cray as part of
* `spack buildcache create -r ...`
* `spack install` when installation is from binary cache

@scottwittenburg @shahzebsiddiqui @becker33

eugeneswalker committed:
make_package_relative: relocate rpaths on cray

eugeneswalker committed:
relocate_package: relocate rpaths on cray

gartung reviewed (approved):
Good catch!

gartung commented:
@tgamblin This is needed to create buildcaches correctly on Cray systems.

gartung commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -619,6 +620,7 @@ def is_backup_file(file):
                                              old_prefix,
                                              new_prefix)
         if (spec.architecture.platform == 'linux' or
+            spec.architecture.platform == 'cray' or
```
Lines  622-625 should probably be contracted to 
```
if (platform.system().lower() == 'linux'):
```

gartung commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -526,6 +526,7 @@ def make_package_relative(workdir, spec, allow_root):
         relocate.make_macho_binaries_relative(cur_path_names, orig_path_names,
                                               old_layout_root)
     if (spec.architecture.platform == 'linux' or
+        spec.architecture.platform == 'cray' or
```
Lines 528-532 should probably be contracted to 
```
if (platform.system().lower() == 'linux'):
```

tgamblin replied:
I tried to clean this up by adding a `binary_formats` attribute to the platform.  @becker33 see what you think.

eugeneswalker replied:
So should I change to `platform.system().lower() == 'linux'` ?

gartung replied:
I would use the changes Todd made.

eugeneswalker replied:
Nevermind. I see what you did.

tgamblin committed:
platforms: add `binary_formats` property

We need to know which binary formats are supported on a platform so we
know which types of relocations to try. This adds a list of binary
formats to the platform and removes a bunch of special cases from
`binary_distribution.py`.

becker33 commented on lib/spack/spack/platforms/test.py:
```diff
@@ -3,12 +3,17 @@
+
+    if platform.system().lower() == 'darwin':
+        binary_formats = ['macho']
+
```
Shouldn't this have an `else` case to set it to `elf`?

becker33 replied:
Oh wait no I see it's defaulted.
